pkg/ssh/metrics.go

```diff
@@ -16,5 +34,65 @@ func newPromMetrics() *promMetrics {
 			},
 			[]string{"exit_code"},
 		),
+		channelsCount: prometheus.NewGauge(
+			prometheus.GaugeOpts{
+				Name:      "ssh_channels",
+				Help:      "Number of active SSH channels",
+				Namespace: "pdc_agent",
+			},
+		),
+		tcpConnectionsCount: prometheus.NewCounterVec(
+			prometheus.CounterOpts{
+				Name:      "tcp_connections_total",
+				Help:      "Number of open TCP connections",
+				Namespace: "pdc_agent",
+			},
+			[]string{"target", "status"},
+		),
+		timeToConnect: metrics.NewNativeHistogram(
+			prometheus.HistogramOpts{
+				Name:      "ssh_time_to_connect_seconds",
+				Help:      "Time spent to establish SSH connection",
+				Namespace: "pdc_agent",
+			}),
+	}
+}
+
+type logMetricsParser struct {
+	m         *promMetrics
+	connStart time.Time
+}
+
+func (p logMetricsParser) parseLogMetrics(msg []byte) {
+	switch {
+	case bytes.Contains(msg, []byte("nchannels")):
+		p.channelsCount(msg)
+	case bytes.Contains(msg, []byte("connected to")):
+		p.tcpConnCount(msg, reConnSuccess, "success")
+	case bytes.Contains(msg, []byte("connect_to")):
+		p.tcpConnCount(msg, reConnFailure, "failure")
+	case bytes.Contains(msg, []byte("This is Grafana Private Datasource Connect!")):
+		p.m.timeToConnect.Observe(time.Since(p.connStart).Seconds())
```

Comment:
When I see the line

```
level=info caller=ssh.go:349 ts=2024-11-04T10:29:31.232049Z msg="connect_to prometheus.demo.do.prometheus.iod: unknown host (nodename nor servname provided, or not known)"
```

in my agent (The address is invalid), I don't see an increment in this metric. Why is that?

Reply:
I think the regex for the failure case might be too narrow.

Reply:
Thanks. I've changed it slightly.